README.md
# FutebolPT
[This project is not actively maintained anymore, since widgets are no longer supported by macOS.]

## About
*FutebolPT* is a macOS widget that provides information about the Portuguese main football championship.
It list the matches and results of all rounds of the championship, as well the current standings.

It uses as source the website of the [Liga Portuguesa de Futebol Profissional (Portuguese League for Professional Football)](https://www.ligaportugal.pt/).


# Links
* [Downloads](https://github.com/rcgoncalves/futebolpt/releases/latest)
* [Changelog](https://github.com/rcgoncalves/futebolpt/blob/master/CHANGELOG.md)


# License
FutebolPT

Copyright (C) 2009-2011 Rui Carlos Gonçalves

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see <[https://www.gnu.org/licenses/](https://www.gnu.org/licenses/)>.


CHANGELOG.md
# Changelog

## [3.0.3] - 2011-09-21
### Changed 
- Minor changes.


## [3.0.2] - 2010-08-14
### Fixed
- Fix bugs.


## [3.0.1] - 2010-08-11
### Changed
- Update source URL.


## [3.0.0] - 2009-09-21
### Added
- Add the ability to reduce the widget size, clicking on its title.
- Add info about TV transmission.


## [2.0.0] - 2009-09-20
### Added
- Add the ability to select a particular round.
- Add matches time.
- Add to the classification the number of played, won, drawn, and lost matches.


## [1.0.0] - 2009-09-05
### Added
- Initial version.


FutebolPT.dcproj/project/widget.wdgt/Parts/parts.js
/* 
 This file was generated by Dashcode and is covered by the 
 license.txt included in the project.  You may edit this file, 
 however it is recommended to first turn off the Dashcode 
 code generator otherwise the changes will be lost. This file
 is for files included by Dashcode directly. It will be replaced
 with an optimized version at deploy time.
*/
var dashcodePartSupport = {
    "core": ["Parts/core/utilities.js", "Parts/core/core/base.js", "Parts/core/core/array-additions.js", "Parts/core/core/array-additions-ie.js", "Parts/core/core/set.js", "Parts/core/core/oop.js", "Parts/core/core/function-additions.js", "Parts/core/core/object-additions.js", "Parts/core/core/regex-additions.js", "Parts/core/core/local.js", "Parts/core/core/Error.js", "Parts/core/core/kvo.js", "Parts/core/core/Bindable.js", "Parts/core/core/SortDescriptor.js", "Parts/core/core/transformers.js", "Parts/core/core/Binding.js", "Parts/core/core/kvo-array.js", "Parts/core/core/kvo-array-operators.js", "Parts/core/core/model.js", "Parts/core/core/string-additions.js", "Parts/core/net/Deferred.js", "Parts/core/net/XHR.js", "Parts/core/controllers/Controller.js", "Parts/core/controllers/SelectionProxy.js", "Parts/core/controllers/ObjectController.js", "Parts/core/controllers/ArrayController.js", "Parts/core/controllers/AjaxController.js", "Parts/core/controllers/ModeledXMLProxy.js", "Parts/core/dom/element.js", "Parts/core/dom/event.js", "Parts/core/dom/element-ie.js", "Parts/core/dom/event-ie.js", "Parts/core/views/view-parts.js", "Parts/core/views/view-core.js", "Parts/core/views/Responder.js", "Parts/core/views/View.js", "Parts/core/views/ViewController.js", "Parts/core/views/ImageView.js", "Parts/core/views/ImageLayout.js", "Parts/core/views/DashcodePart.js", "Parts/core/views/FormControl.js", "Parts/core/views/TextField.js", "Parts/core/views/ListView.js", "Parts/core/views/SelectField.js", "Parts/core/views/ToggleButton.js", "Parts/core/views/SearchField.js", "Parts/core/views/Slider.js", "Parts/core/views/EventLoop.js", "Parts/core/views/Page.js", "Parts/core/views/Media.js", "Parts/core/views/Video.js", "Parts/core/views/VideoLayout.js", "Parts/core/views/VideoLegacy.js"],
    "scripts": ["Parts/setup.js", "Parts/datasources.js", "Parts/GlassButton.js", "Parts/InfoButton.js", "Parts/Button.js", "Parts/PopupButton.js"]
};

(function() {
    var scripts = dashcodePartSupport['core'];

    scripts = scripts.concat(dashcodePartSupport['scripts']);

    for(var index in scripts) {
        var path = scripts[index];
        var scriptTag = '<script apple-no-regeneration="yes" type="text/javascript" src="' + path + '"></script>';
        
        document.write( scriptTag );
    }
    
})();


FutebolPT.dcproj/project/widget.wdgt/Parts/setup.js
/* 
 This file was generated by Dashcode and is covered by the 
 license.txt included in the project.  You may edit this file, 
 however it is recommended to first turn off the Dashcode 
 code generator otherwise the changes will be lost.
 */
var dashcodePartSpecs = {
    "classification": { "creationFunction": "CreateButton", "leftImageWidth": 1, "onclick": "showClassification", "rightImageWidth": 10, "text": "Classificação", "view": "DC.View" },
    "done": { "creationFunction": "CreateGlassButton", "onclick": "showFront", "text": "Done", "view": "DC.View" },
    "info": { "backgroundStyle": "black", "creationFunction": "CreateInfoButton", "foregroundStyle": "white", "frontID": "front", "onclick": "showBack", "view": "DC.View" },
    "popup": { "creationFunction": "CreatePopupButton", "leftImageWidth": 10, "name": "round", "rightImageWidth": 16, "view": "DC.View" },
    "results": { "creationFunction": "CreateButton", "leftImageWidth": 10, "onclick": "showResults", "rightImageWidth": 1, "text": "Resultados", "view": "DC.View" },
    "text1": { "text": "FutebolPT 3.0.3", "view": "DC.Text" },
    "text2": { "text": "Desenvolvido por:", "view": "DC.Text" },
    "text3": { "text": "rcgoncalves.pt@gmail.com", "view": "DC.Text" },
    "text5": { "text": "Rui Carlos A. Gonçalves", "view": "DC.Text" },
    "text6": { "text": "Licença: GPLv2", "view": "DC.Text" },
    "text8": { "text": "Jornada nº", "view": "DC.Text" },
    "title": { "text": "FutebolPT", "view": "DC.Text" }
};





FutebolPT.dcproj/project/widget.wdgt/Scripts/Parser.js
/*
 * Copyright (C) 2009, Rui Carlos Gonçalves (rcgoncalves.pt@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

/**
 * Returns the first element from XML formatted string 'str', with tag name 'tag'.
 * It starts searching from 'start'.
 */
function getFirstElementByTagName(str,tag,start)
{
  var strx=str;
  if(start) strx=strx.substr(start);
  strx=removeComments(strx);
  var res="";
  var open=new RegExp("<"+tag+"[^>]*>","i");
  var close=new RegExp("</"+tag+" *>","i");
  var pos1=strx.search(open);
  var pos2;
  var pos=pos1;
  var count;
  
  if(pos<0) return "";
  else
  {
    strx=strx.substr(pos);
    res=strx.substr(0,1);
    strx=strx.substr(1);
    count=1;
    var error=false;
    
    while(count>0 && !error)
    {
      pos1=strx.search(open);
      pos2=strx.search(close);
      if(pos2<0) error=true;
      
      if(pos1<pos2 && pos1>=0) //open tag
      {
        count++;
        pos=pos1;
      }
      else if(pos2>=0) //close tag
      {
        count--;
        pos=pos2;
      }
      
      res+=strx.substr(0,pos+1);
      strx=strx.substr(pos+1);
    }
    
    if(error) return "";
    else return res+"/"+tag+">";
  }
}

/**
 * Remove XML comments from the string 'str'.
 */
function removeComments(str)
{
  var start=str.indexOf("<!--");
  var end;
  while(start>=0)
  {
    end=str.indexOf("-->");
    if(end<=start) start=-1;
    else
    {
      str=str.substr(0,start)+str.substr(end+3);
      start=str.indexOf("<!--");
    }
  }
  
  return str;
}

function removeSpaces(str)
{
  var res=str.replace(/\s+/g,"");
  return res;
}

function removeTags(str)
{
  var res=str.replace(/<[^>]*>/g,"");
  return res;
}

FutebolPT.dcproj/project/widget.wdgt/main.js
/*
 * Copyright (C) 2009, Rui Carlos Gonçalves (rcgoncalves.pt@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

//
// Function: load()
// Called by HTML body element's onload event when the widget is ready to start
//
function load()
{
    dashcode.setupParts();
    _compress();
    populateRounds();
}

//
// Function: remove()
// Called when the widget has been removed from the Dashboard
//
function remove()
{
    // Stop any timers to prevent CPU usage
    // Remove any preferences as needed
    // widget.setPreferenceForKey(null, dashcode.createInstancePreferenceKey("your-key"));
}

//
// Function: hide()
// Called when the widget has been hidden
//
function hide()
{
    // Stop any timers to prevent CPU usage
}

//
// Function: show()
// Called when the widget has been shown
//
function show()
{
    // Restart any timers that were stopped on hide
}

//
// Function: sync()
// Called when the widget has been synchronized with .Mac
//
function sync()
{
    // Retrieve any preference values that you need to be synchronized here
    // Use this for an instance key's value:
    // instancePreferenceValue = widget.preferenceForKey(null, dashcode.createInstancePreferenceKey("your-key"));
    //
    // Or this for global key's value:
    // globalPreferenceValue = widget.preferenceForKey(null, "your-key");
}

//
// Function: showBack(event)
// Called when the info button is clicked to show the back of the widget
//
// event: onClick event from the info button
//
function showBack(event)
{
    var front = document.getElementById("front");
    var back = document.getElementById("back");

    if (window.widget) {
        widget.prepareForTransition("ToBack");
    }

    front.style.display = "none";
    back.style.display = "block";

    if (window.widget) {
        setTimeout('widget.performTransition();', 0);
    }
}

//
// Function: showFront(event)
// Called when the done button is clicked from the back of the widget
//
// event: onClick event from the done button
//
function showFront(event)
{
    var front = document.getElementById("front");
    var back = document.getElementById("back");

    if (window.widget) {
        widget.prepareForTransition("ToFront");
    }

    front.style.display="block";
    back.style.display="none";

    if (window.widget) {
        setTimeout('widget.performTransition();', 0);
    }
}

if (window.widget) {
    widget.onremove = remove;
    widget.onhide = hide;
    widget.onshow = show;
    widget.onsync = sync;
}

function showResults(event)
{    
  var feedURL='http://www.lpfp.pt/liga_zon_sagres/Pages/Jornada.aspx';
  var onloadHandler=function() {processResults(xmlRequest);};

  var xmlRequest = new XMLHttpRequest();
  xmlRequest.onload = onloadHandler;
  xmlRequest.open('GET', feedURL);
  xmlRequest.setRequestHeader('Cache-Control', 'no-cache');
  xmlRequest.send(null);
}

function processResults(xmlRequest)
{
  window.resizeTo(320, 380);
  $('text').style.visibility='visible';
  $('info').style.visibility='visible';
  $('front').childNodes[1].setAttribute('src','Parts/Images/front.png');
  
  if (xmlRequest.status == 200)
  {
    var html=xmlRequest.responseText;
    $('text').innerHTML='';
    $('tmp').innerHTML=html;
    var results=$('MatchDayWP').innerHTML;
    $('tmp').innerHTML='';
    results=getFirstElementByTagName(results,'div',1);
    var round=(results.match(/Jornada\s*nº\s*\d+/))[0];
    round=(round.match(/\d+/))[0];
    results=results.replace(/<h3>Jornada\s*nº\s*\d+<\/h3>/,'<h4>&nbsp;</h4>');
    results=results.replace(/width=\"\d+px\"/,'width="100%"');//"
    results=results.replace(/src=\"([^\"]*)\"/g,"src=\"http://www.lpfp.pt/$1\"");//"
    $('tmp').innerHTML=results;
    var contents=$('tg').childNodes[0];
    var len=contents.childNodes.length;

    var node=contents.childNodes[0];
    node.removeChild(node.childNodes[0]);
    node.removeChild(node.childNodes[1]);
    for(var i=2;i<7;i++)
    {
      node.removeChild(node.childNodes[2]);
    }

    for(var i=2;i<len;i+=2)
    {
      var node=contents.childNodes[i];
      node.childNodes[0].innerHTML=node.childNodes[1].innerHTML;
      node.childNodes[0].setAttribute('class','club_l');
      node.childNodes[1].innerHTML=node.childNodes[2].innerHTML;
      node.childNodes[1].setAttribute('class','crest_l');
      node.childNodes[2].innerHTML=node.childNodes[3].innerHTML;
      node.childNodes[2].setAttribute('class','result');
      node.childNodes[3].innerHTML=node.childNodes[4].innerHTML;
      node.childNodes[3].setAttribute('class','crest_r');
      node.childNodes[4].innerHTML=node.childNodes[5].innerHTML;
      node.childNodes[4].setAttribute('class','club_r');
      node.childNodes[5].innerHTML=node.childNodes[7].innerText+'<br/>'+node.childNodes[9].innerText;
      var className=node.childNodes[11].childNodes[0].className;
      if(className)
      {
        if(className.match('sporttv1'))
        {
          node.childNodes[5].setAttribute('title','SportTV1');
        }
        else if(className.match('tvi'))
        {
          node.childNodes[5].setAttribute('title','TVI');
        }
      }
      node.childNodes[5].setAttribute('class','date');
      
      for(var j=6;j<13;j++)
      {
        node.removeChild(node.childNodes[6]);
      }
    }
  
    $('tg').innerHTML=contents.innerHTML;
    results=$('tmp').innerHTML;
    $('tmp').innerHTML='';
    text.innerHTML='<div id="results_contents">'+results+'</div>';
    
    var cb=$('popup');
    cb.childNodes[0].childNodes[1].innerHTML=round;
    cb.childNodes[2].selectedIndex=round-1;
    $('text8').style.visibility='visible';
    $('popup').style.visibility='visible';
  }
  else
  {
    alert('Error fetching data: HTTP status ' + xmlRequest.status);
    $('text').innerHTML="Não foi possível obter os dados!\n(HTTP status "+xmlRequest.status+')';
  }
}

function showClassification(event)
{
  var feedURL='http://www.lpfp.pt/liga_zon_sagres/Pages/Jornada.aspx';
  var onloadHandler=function() {processClassification(xmlRequest);};

  var xmlRequest = new XMLHttpRequest();
  xmlRequest.onload = onloadHandler;
  xmlRequest.open('GET', feedURL);
  xmlRequest.setRequestHeader('Cache-Control','no-cache');
  xmlRequest.send(null);
}

function processClassification(xmlRequest)
{
  window.resizeTo(320, 380);
  $('text').style.visibility='visible';
  $('info').style.visibility='visible';
  $('front').childNodes[1].setAttribute('src','Parts/Images/front.png');
  
	if (xmlRequest.status == 200)
  {
    $('text8').style.visibility='hidden';
    $('popup').style.visibility='hidden';
    
    var html=xmlRequest.responseText;
    $('tmp').innerHTML=html;
    var classification=$('sidebar-nav-comp-display').innerHTML;
    $('tmp').innerHTML='';
    classification=getFirstElementByTagName(classification,'table',0);
    classification=getFirstElementByTagName(classification,'table',1);
    classification=classification.replace(/<img[^>]*>/g,'');
    $('text').innerHTML='<div id="classification_contents">'+classification+'</div>';
	}
	else
  {
		alert('Error fetching data: HTTP status ' + xmlRequest.status);
    $('text').innerHTML="Não foi possível obter os dados!\n(HTTP status "+xmlRequest.status+')';
	}
}

/*function processClassification(xmlRequest)
{
	if (xmlRequest.status == 200)
  {
    $('text8').style.visibility='hidden';
    $('popup').style.visibility='hidden';
    
    var html=xmlRequest.responseText;
    $('tmp').innerHTML=html;
    var results=$('MatchDayWP').innerHTML;
    $('tmp').innerHTML='';
    classification=getFirstElementByTagName(results,'div',0);
    classification=classification.replace(/<img[^>]*>/g,'');
    classification=classification.replace(/<h3>.*<\/h3>/,'');
    classification=classification.replace(/width=\"\d+px\"/,'width="100%"');//"
    $('text').innerHTML=classification;
	}
	else
  {
		alert('Error fetching data: HTTP status ' + xmlRequest.status);
    $('text').innerHTML="Não foi possível obter os dados!\n(HTTP status "+xmlRequest.status+')';
	}
}*/

function showRound(event)
{
  var round=$('popup').childNodes[2].selectedIndex+1;
  var feedURL='http://www.lpfp.pt/liga_zon_sagres/Pages/Jornada.aspx?Jornada='+round;
  var onloadHandler=function() {processResults(xmlRequest);};

  var xmlRequest = new XMLHttpRequest();
  xmlRequest.onload = onloadHandler;
  xmlRequest.open('GET', feedURL);
  xmlRequest.setRequestHeader('Cache-Control', 'no-cache');
  xmlRequest.send(null);
}

function populateRounds()
{	
  var nrounds=30;
  var rounds=$('popup').childNodes[2];
	while (rounds.hasChildNodes())
		rounds.removeChild(round.firstChild);
	
  for(var i=1; i<=nrounds; i++)
  {
    var element=document.createElement('option');
    element.innerText=i;
    element.setAttribute('name',i);
    element.selected=false;
    rounds.appendChild(element);
  }
}

function $(id)
{
  return document.getElementById(id);
}

function compress(event)
{
  _compress();
}

function _compress()
{
  $('text8').style.visibility='hidden';
  $('popup').style.visibility='hidden';
  $('text').style.visibility='hidden';
  $('info').style.visibility='hidden';
  var elem=$('front').childNodes;
  $('front').childNodes[1].setAttribute('src','Parts/Images/front_small.png');
  window.resizeTo(320, 60);
}


function openProjectPage(event)
{
    if( window.widget) {
        widget.openURL( "http://rcgoncalves.pt/project/futebolpt/" );
        return false;
    }
}


function openRCG(event)
{
    if( window.widget) {
        widget.openURL( "http://rcgoncalves.pt/" );
        return false;
    }
}
